code/common/grafana.q

```diff
@@ -0,0 +1,146 @@
+// user defined column name of time column
+.gkdb.timeCol:@[value;`.gdkb.timeCol;`time];
+// user defined column name of sym column
+.gkdb.sym:@[value;`.gkdb.sym;`sym];
+// user defined date range to find syms from
+.gkdb.timeBackdate:@[value;`.gkdb.timeBackdate;2D];
+// user defined number of ticks to return
+.gkdb.ticks:@[value;`.gkdb.ticks;1000];
+
+// json types of kdb datatypes
+.gkdb.types:.Q.t!`array`boolean,(3#`null),(5#`number),11#`string;
+// milliseconds between 1970 and 2000
+.gkdb.epoch:946684800000;
+
+// wrapper if user has custom .z.pp
+.z.pp:{[f;x]f x;zpp x}[@[value;`.z.pp;{{[x]}}]];
+
+// return alive response for GET requests
+.z.ph:{[f;x]"HTTP/1.1 200 OK\r\nConnection: close\r\n\r\n";f x}[@[value;`.z.ph;{{[x]}}]];
+
+// retrieve and convert Grafana HTTP POST request then process as either timeseries or table
+zpp:{
+  // get API url from request
+  r:" " vs first x;
+  // convert grafana mesage to q
+  rqt:.j.k r 1;
+  $["query"~r 0;query[rqt];"search"~r 0;search rqt;`$"Annotation url nyi"]
+ };
+
+query:{[rqt]
+  // retrieve final query and append to table to log
+  rqtype:raze rqt[`targets]`type;
+  :.h.hy[`json]$[rqtype~"timeserie";tsfunc rqt;tbfunc rqt];
+ };
+
+finddistinctsyms:{?[x;enlist(>;.gkdb.timeCol;(-;.z.p;.gkdb.timeBackdate));1b;{x!x}enlist .gkdb.sym].gkdb.sym};
+
+search:{[rqt]
+  // build drop down case options from tables in port
+  tabs:tables[];
+  symtabs:?[.gkdb.sym in'cols each tabs;tabs;count[tabs]#`] except `;
+  timetabs:?[.gkdb.timeCol in'cols each tabs;tabs;count[tabs]#`] except `;
+  rsp:string tabs;
+  if[count timetabs;
+    rsp,:s1:string` sv/:`t,/:timetabs;
+    rsp,:s2:string` sv/:`g,/:timetabs; 
+    rsp,:raze(s2,'"."),/:'c1:string {(cols x) where`number=.gkdb.types (0!meta x)`t}each timetabs;
+    rsp,:raze((string` sv/:`o,/:timetabs),'"."),/:'c1;
+    if[count symtabs;show finddistinctsyms'[timetabs];
+      rsp,:raze(s1,'"."),/:'c2:string each finddistinctsyms'[timetabs];
+      rsp,:raze((string` sv/:`o,/:timetabs),'"."),/:'{x[0] cross ".",'string finddistinctsyms x 1}each (enlist each c1),'timetabs;
+     ];
+   ];
+  :.h.hy[`json].j.j rsp;
+ };
+
+diskvals:{c:count[x]- .gkdb.ticks+til .gkdb.ticks;get'[.Q.ind[x;c]]};
+memvals:{get'[?[x;enlist(within;`i;count[x]-(.gkdb.ticks),0);0b;()]]};
+catchvals:{@[diskvals;x;memvals x]};
```

Comment:
This will also run `memvals` regardless of whether `diskvals` succeeds or not. Need to do something like wrap it in a function to ensure it is only executed if there is an error.
```
catchvals:{@[diskvals;x;{[x;y]memvals x}[x]]};
```